### ISSUE TITLE
bug with index() matching beyond end of string when \0 bytes (00000000) are involved

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9315]
Migrated from [rt.perl.org#53746](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=53746) (status was 'resolved')

Searchable as RT53746$

**Comment by p5pRT** [comment]
### From insana@mediaglyphs.org

This is a bug report for perl from insana@&#8203;ebi\.ac\.uk\,
generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.

\# by Giuseppe Insana \<insana@&#8203;ebi\.ac\.uk>
\# May 5 2008

$msg="aaZaa";
$string1="aa";
$string2="aaa";
print "msg \(bits\)&#8203;: "\,unpack\('b\*'\,$msg\)\,"\\n";
print "indexing string1 \($string1\) and string2 \($string2\) inside msg 
\($msg\)\\n";
print index\($msg\,$string1\,0\)\,"\\n";
print index\($msg\,$string2\,0\)\,"\\n";

$msg="\\0\\0Z\\0\\0";
$string1="\\0\\0";
$string2="\\0\\0\\0";
print "msg \(bits\)&#8203;: "\,unpack\('b\*'\,$msg\)\,"\\n";
print "indexing string1 \($string1\) and string2 \($string2\) inside msg 
\($msg\)\\n";
print index\($msg\,$string1\,0\)\,"\\n";
print index\($msg\,$string2\,0\)\,"\\n";

print "\\n result is different\, but actually for string2 index should 
always return \-1 in both cases\, shouldn't it??\\n\\n";



---
Flags&#8203;:
&nbsp;     category=core
&nbsp;     severity=medium


---
Site configuration information for perl 5\.10\.0&#8203;:

Configured by SYSTEM at Thu Jan 10 11&#8203;:00&#8203;:30 2008\.

Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;   Platform&#8203;:
&nbsp;     osname=MSWin32\, osvers=5\.00\, archname=MSWin32\-x86\-multi\-thread
&nbsp;     uname=''
&nbsp;     config\_args='undef'
&nbsp;     hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;     useithreads=define\, usemultiplicity=define
&nbsp;     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;     use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;     usemymalloc=n\, bincompat5005=undef
&nbsp;   Compiler&#8203;:
&nbsp;     cc='cl'\, ccflags ='\-nologo \-GF \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32 
\-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT \-DUSE\_SITECUSTOMIZE 
\-DPRIVLIB\_LAST\_IN\_INC \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS 
\-DUSE\_PERLIO \-DPERL\_MSVCRT\_READFIX'\,
&nbsp;     optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
&nbsp;     cppflags='\-DWIN32'
&nbsp;     ccversion='12\.00\.8804'\, gccversion=''\, gccosandvers=''
&nbsp;     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;     d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, 
Off\_t='\_\_int64'\, lseeksize=8
&nbsp;     alignbytes=8\, prototype=define
&nbsp;   Linker and Libraries&#8203;:
&nbsp;     ld='link'\, ldflags ='\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf 
\-libpath&#8203;:"D&#8203;:\\apps\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;     libpth=\\lib
&nbsp;     libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib 
&nbsp;  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib 
netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib  version\.lib 
odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;     perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib 
winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib 
oleaut32\.lib  netapi32\.lib uuid\.lib ws2\_32\.lib mpr\.lib winmm\.lib 
version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;     libc=msvcrt\.lib\, so=dll\, useshrplib=true\, libperl=perl510\.lib
&nbsp;     gnulibc\_version=''
&nbsp;   Dynamic Linking&#8203;:
&nbsp;     dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;     cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-debug 
\-opt&#8203;:ref\,icf  \-libpath&#8203;:"D&#8203;:\\apps\\Perl\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:
&nbsp;     ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY
&nbsp;     32809 Load 'loadable object' with non\-default file extension
&nbsp;     32728 64\-bit fix for Time&#8203;::Local



---
@&#8203;INC for perl 5\.10\.0&#8203;:
&nbsp;     D&#8203;:/apps/Perl/site/lib
&nbsp;     D&#8203;:/apps/Perl/lib
&nbsp;     \.



---
Environment for perl 5\.10\.0&#8203;:
&nbsp;     HOME=D&#8203;:\\HOME
&nbsp;     LANG \(unset\)
&nbsp;     LANGUAGE \(unset\)
&nbsp;     LD\_LIBRARY\_PATH \(unset\)
&nbsp;     LOGDIR \(unset\)
 
PATH=D&#8203;:\\apps\\Perl\\site\\bin;D&#8203;:\\apps\\Perl\\bin;D&#8203;:\\apps\\PHP\\;C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;C&#8203;:\\Program 
Files\\GNU\\GnuPG;D&#8203;:\\apps\\cvsnt;C&#8203;:\\USR\\BIN;D&#8203;:\\apps\\Python25;
&nbsp;     PERL\_BADLANG \(unset\)
&nbsp;     SHELL=/bin/sh

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

Easier example&#8203;:

perl5\.8\.8 \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
\# Output&#8203;: \-1

perl5\.10\.0 \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
\# Output&#8203;: 3

perl\-blead \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
\# Output&#8203;: 3


Note that&#8203;:

perl\-blead \-wle 'print index\("\\0\\0Z\\0\\0"\, "\\0\\0\\0"\, 0\);'
\# Output&#8203;: \-1

perl\-blead \-wle 'my $s = "\\0\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
\# Output&#8203;: \-1

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @abigail

On Mon\, May 05\, 2008 at 11&#8203;:50&#8203;:44AM \-0700\, Bram via RT wrote&#8203;:
> Easier example&#8203;:
> 
> perl5\.8\.8 \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
> \# Output&#8203;: \-1
> 
> perl5\.10\.0 \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
> \# Output&#8203;: 3
> 
> perl\-blead \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
> \# Output&#8203;: 3


Attached is a set of tests\.


Abigail

**Comment by p5pRT** [comment]
### From @abigail

<details><summary>Inline Patch</summary>

```diff
--- t/op/index.t.orig	2008-05-06 14:08:20.000000000 +0200
+++ t/op/index.t	2008-05-06 14:56:41.000000000 +0200
@@ -7,7 +7,7 @@
 }
 
 use strict;
-plan( tests => 69 );
+plan( tests => 111 );
 
 run_tests() unless caller;
 
@@ -160,4 +160,43 @@
     is(index($t, 'xyz'), 4, "0xfffffffd and utf8cache");
 }
 
+
+# Tests for NUL characters.
+{
+    my @tests = (
+        ["",            -1, -1, -1],
+        ["foo",         -1, -1, -1],
+        ["\0",           0, -1, -1],
+        ["\0\0",         0,  0, -1],
+        ["\0\0\0",       0,  0,  0],
+        ["foo\0",        3, -1, -1],
+        ["foo\0foo\0\0", 3,  7, -1],
+    );
+    foreach my $l (1 .. 3) {
+        my $q = "\0" x $l;
+        my $i = 0;
+        foreach my $test (@tests) {
+            $i ++;
+            my $str = $$test [0];
+            my $res = $$test [$l];
+
+            {
+                local $::TODO = ($l == 3 && $i == 7 ? "Bug #53746" : "");
+                is (index ($str, $q), $res, "Find NUL character(s)");
+            }
+
+            #
+            # Bug #53746 shows a difference between variables and literals,
+            # so test literals as well.
+            #
+            my $test_str = qq {is (index ("$str", "$q"), $res, } .
+                           qq {"Find NUL character(s)")};
+               $test_str =~ s/\0/\\0/g;
+
+            eval $test_str;
+            die $@ if $@;
+        }
+    }
+}
+
 }
```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, May 06\, 2008 at 02&#8203;:57&#8203;:36PM \+0200\, Abigail wrote&#8203;:
> On Mon\, May 05\, 2008 at 11&#8203;:50&#8203;:44AM \-0700\, Bram via RT wrote&#8203;:
> > Easier example&#8203;:
> > 
> > perl5\.8\.8 \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
> > \# Output&#8203;: \-1
> > 
> > perl5\.10\.0 \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
> > \# Output&#8203;: 3
> > 
> > perl\-blead \-wle 'my $s = "\\0\\0\\0"; print index\("\\0\\0Z\\0\\0"\, $s\, 0\);'
> > \# Output&#8203;: 3
> 
> 
> Attached is a set of tests\.

Thanks\, applied as change \#33951\.
I've fixed the bug itself as change \#33952\.
\(Whoo hoo\! A bug that was easy to locate\, fix\, and test for\. You don't
see that very often :\-\)\.

\-\- 
Overhead\, without any fuss\, the stars were going out\.
&nbsp;    \-\- Arthur C Clarke

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
